src/shared.c

```diff
@@ -602,14 +603,311 @@ void hc_string_trim_trailing (char *s)
   s[new_len] = 0;
 }
 
-size_t hc_fread (void *ptr, size_t size, size_t nmemb, FILE *stream)
+#if defined (__CYGWIN__)
+// workaround for zlib with cygwin build
+int _wopen(const char *path, int oflag, ...)
+{
+  va_list ap;
+  va_start (ap, oflag);
+  int r = open (path, oflag, ap);
+  va_end (ap);
+  return r;
+}
+#endif
+
+bool hc_fopen (HCFILE *fp, const char *path, char *mode, int file_format)
+{
+  if (!path || !mode) return false;
+
+  int oflag = -1;
+
+  int fmode = S_IRUSR|S_IWUSR;
+
+  if (!strncmp (mode, "a", 1) || !strncmp (mode, "ab", 2))
+  {
+    oflag = O_WRONLY | O_CREAT | O_APPEND;
+
+    #if defined(MSDOS) || defined(OS2) || defined(WIN32) || defined(_WIN32) || defined(__CYGWIN__)
+    if (!strncmp (mode, "ab", 2)) oflag |= O_BINARY;
+    #endif
+  }
+  else if (!strncmp (mode, "r", 1) || !strncmp (mode, "rb", 2))
+  {
+    oflag = O_RDONLY;
+    fmode = -1;
+
+    #if defined(MSDOS) || defined(OS2) || defined(WIN32) || defined(_WIN32) || defined(__CYGWIN__)
+    if (!strncmp (mode, "rb", 2)) oflag |= O_BINARY;
+    #endif
+  }
+  else if (!strncmp (mode, "w", 1) || !strncmp (mode, "wb", 2))
+  {
+    oflag = O_WRONLY | O_CREAT | O_TRUNC;
+
+    #if defined(MSDOS) || defined(OS2) || defined(WIN32) || defined(_WIN32) || defined(__CYGWIN__)
+    if (!strncmp (mode, "wb", 2)) oflag |= O_BINARY;
+    #endif
+  }
+  else
+  {
+    // ADD more strncmp to handle more "mode"
+    return false;
+  }
+
+  if (file_format == HCFILE_FORMAT_PLAIN)
+  {
+    unsigned char check[3] = { 0 };
+
+    int fd_tmp = open (path, O_RDONLY);
+
+    lseek (fd_tmp, 0, SEEK_SET);
+
+    size_t s = read (fd_tmp, check, sizeof(check));
+
+    if (s == 3 && (check[0] == 0x1f && check[1] == 0x8b && check[2] == 0x08)) file_format = HCFILE_FORMAT_GZIP;
+
+    close (fd_tmp);
+  }
+
+  fp->fd = (fmode == -1) ? open (path, oflag) : open (path, oflag, fmode);
+
+  if (fp->fd == -1) return false;
+
+  if (file_format == HCFILE_FORMAT_PLAIN)
+  {
+    if (!(fp->pfp = fdopen (fp->fd, mode)))  return false;
+
+    fp->is_gzip = false;
+  }
+  else // HCFILE_FORMAT_GZIP
+  {
+    if (!(fp->gfp = gzdopen (fp->fd, mode))) return false;
+
+    fp->is_gzip = true;
+    fp->pfp     = NULL;
+  }
+
+  fp->path = path;
+  fp->mode = mode;
+
+  return true;
+}
+
+size_t hc_fread (void *ptr, size_t size, size_t nmemb, HCFILE *fp)
 {
-  return fread (ptr, size, nmemb, stream);
+  size_t n = -1;
+
+  if (fp == NULL) return n;
+
+  if (fp->is_gzip)
```

Comment:
Please fix missing { and } blocks. It's easier to read.

Reply:
You mean for "if (fp->is_gzip)" ?

Reply:
Sorry what I mean is make this:

```
  if (fp->is_gzip)
    gzrewind (fp->gfp);
  else
    rewind (fp->pfp);
```

like this:

```  
  if (fp->is_gzip)
  {
    gzrewind (fp->gfp);
  }
  else
  {
    rewind (fp->pfp);
  }

```